### PR TITLE
proc_macro/bridge: stop using a remote object handle for proc_macro Ident and Literal

### PR DESCRIPTION
**Pull request by mystor**
This is the fourth part of https://github.com/rust-lang/rust/pull/86822, split off as requested in https://github.com/rust-lang/rust/pull/86822#pullrequestreview-1008655452. This patch transforms the `Ident` and `Group` types into structs serialized over IPC rather than handles.

Symbol values are interned on both the client and server when deserializing, to avoid unnecessary string copies and keep the size of `TokenTree` down. To do the interning efficiently on the client, the proc-macro crate is given a vendored version of the fxhash hasher, as `SipHash` appeared to cause performance issues. This was done rather than depending on `rustc_hash` as it is unfortunately difficult to depend on crates from within `proc_macro` due to it being built at the same time as `std`.

In addition, a custom arena allocator and symbol store was also added, inspired by those in `rustc_arena` and `rustc_span`. To prevent symbol re-use across multiple invocations of a macro on the same thread, a new range of `Symbol` names are used for each invocation of the macro, and symbols from previous invocations are cleaned-up.

In order to keep `Ident` creation efficient, a special ASCII-only case was added to perform ident validation without using RPC for simple identifiers. Full identifier validation couldn't be easily added, as it would require depending on the `rustc_lexer` and `unicode-normalization` crates from within `proc_macro`. Unicode identifiers are validated and normalized using RPC.

See the individual commit messages for more details on trade-offs and design decisions behind these patches.

### CODE REVIEW

**Comment by mystor** [comment]
Interesting - I wonder if my local measurements of perf improvements from changing to using `Rc<str>` were inaccurate and it's actually faster to continue to use interned strings...

File: library/proc_macro/src/bridge/client.rs [hunk 1]
```diff
@@ -277,6 +338,12 @@ struct Bridge<'a> {
     /// Server-side function that the client uses to make requests.
     dispatch: closure::Closure<'a, Buffer, Buffer>,
 
+    /// Server-side function to validate and normalize an ident.
+    validate_ident: extern "C" fn(buffer::Slice<'_>, &mut Buffer) -> bool,
+
+    /// Interned store for storing symbols within the client.
+    symbols: handle::InternedStore<Rc<str>>,
```
Comment (eddyb) [inline library/proc_macro/src/bridge/client.rs:345]: Random thought: since this uses `HashMap` without the `Fx` hash the compiler uses, it might be significantly slower to hash? Maybe switching hashers can help (though if it can help with the existing handle stores, we might want to land that as a separate PR).

**Comment by mystor** [comment]
Tried out a few changes in the most recent version (doing an optimized ASCII-only ident validation within the client, and using a fancier interning solution with an arena). Not sure if this will improve the situation and the code (especially for the arena) is pretty unsafe (as `proc_macro` can't depend on anything), so it probably needs some clean-up if this fixes the issues.

File: library/proc_macro/src/lib.rs [hunk 2]
```diff
@@ -1333,7 +1379,47 @@ impl Literal {
     // was 'c' or whether it was '\u{63}'.
     #[unstable(feature = "proc_macro_span", issue = "54725")]
     pub fn subspan<R: RangeBounds<usize>>(&self, range: R) -> Option<Span> {
-        self.0.subspan(range.start_bound().cloned(), range.end_bound().cloned()).map(Span)
+        bridge::client::FreeFunctions::literal_subspan(
```
Comment (eddyb) [inline library/proc_macro/src/lib.rs:1382]: Shouldn't this be a "static method" on `Span`?

File: library/proc_macro/src/lib.rs [hunk 3]
```diff
@@ -1333,7 +1379,47 @@ impl Literal {
...
+            self.0.clone(),
+            range.start_bound().cloned(),
+            range.end_bound().cloned(),
+        )
+        .map(Span)
+    }
+
+    fn with_symbol_suffix<R>(&self, f: impl FnOnce(&str, &str) -> R) -> R {
```
Comment (eddyb) [inline library/proc_macro/src/lib.rs:1390]:
```suggestion
    fn with_symbol_and_suffix<R>(&self, f: impl FnOnce(&str, &str) -> R) -> R {
```

File: library/proc_macro/src/lib.rs [hunk 4]
```diff
@@ -1333,7 +1379,47 @@ impl Literal {
...
+        self.0.symbol.with(|symbol| match self.0.suffix {
+            Some(suffix) => suffix.with(|suffix| f(symbol, suffix)),
+            None => f(symbol, ""),
+        })
+    }
+
+    fn stringify_impl<R>(&self, f: impl FnOnce(&[&str]) -> R) -> R {
```
Comment (eddyb) [inline library/proc_macro/src/lib.rs:1397]:
```suggestion
    fn with_stringify_parts<R>(&self, f: impl FnOnce(&[&str]) -> R) -> R {
```
(and/or add a comment explaining the callback and that this is like this for performance reasons)

File: library/proc_macro/src/bridge/symbol.rs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+/// Basic interner for a `Symbol`, inspired by the one in `rustc_span`.
+struct Interner {
+    arena: arena::Arena,
+    // SAFETY: These `'static` lifetimes are actually references to data owned
+    // by the Arena. This is safe, as we never return them as static references
+    // from `Interner`.
+    names: fxhash::FxHashMap<&'static str, Symbol>,
+    strings: Vec<&'static str>,
```
Comment (eddyb) [inline library/proc_macro/src/bridge/symbol.rs:151]:
```suggestion
    str_to_sym: fxhash::FxHashMap<&'static str, Symbol>,
    sym_to_str: Vec<&'static str>,
```
(one cool trick is that if you had access to the `indexmap` crate, this could be just a single `IndexSet<&'static str>`... alas)

Comment (eddyb) [reply library/proc_macro/src/bridge/symbol.rs:151]: Oh I see, there's "name" used elsewhere in this module. Frankly I think that's confusing, and either "sym(bol)" should be used, or "index" where it benefits to be explicit.

File: library/proc_macro/src/bridge/symbol.rs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    // The offset to apply to symbol names stored in the interner. This is used
+    // to ensure that symbol names are not re-used after the interner is
+    // cleared.
+    offset: u32,
```
Comment (eddyb) [inline library/proc_macro/src/bridge/symbol.rs:155]: `first_valid_sym`/`first_valid_index`, or using "base" somehow, might be better?
Also, should this be `NonZeroU32`?

File: library/proc_macro/src/bridge/server.rs [hunk 7]
```diff
@@ -57,6 +68,7 @@ macro_rules! define_mark_types_impls {
     }),* $(,)?) => {
         impl<S: Types> Types for MarkedTypes<S> {
             $(type $name = Marked<S::$name, client::$name>;)*
+            type Symbol = Marked<S::Symbol, client::Symbol>;
```
Comment (eddyb) [inline library/proc_macro/src/bridge/server.rs:71]: Could it be added to the main RPC API macro, but with no methods? (I forget what else that might generate)

File: library/proc_macro/src/bridge/handle.rs [hunk 8]
```diff
@@ -51,31 +53,15 @@ impl<T> IndexMut<Handle> for OwnedStore<T> {
...
-    fn build_hasher(&self) -> Self::Hasher {
-        Self::Hasher::new()
-    }
-}
-
 /// Like `OwnedStore`, but avoids storing any value more than once.
 pub(super) struct InternedStore<T: 'static> {
     owned: OwnedStore<T>,
-    interner: HashMap<T, Handle, NonRandomState>,
+    interner: FxHashMap<T, Handle>,
```
Comment (eddyb) [inline library/proc_macro/src/bridge/handle.rs:59]: Two thoughts here:
* any chance this is significant enough on its own to get tested in a separate PR?
* just realized that the "base index" trick could work here to allow a simple `Vec` backing storage - I guess the handle counters would need to become thread-local in that case?

File: library/proc_macro/src/bridge/client.rs [hunk 9]
```diff
@@ -430,6 +411,10 @@ fn run_client<A: for<'a, 's> DecodeMut<'a, 's, ()>, R: Encode<()>>(
                 // at the moment, so this is also potentially preventing UB).
                 buf.clear();
                 Ok::<_, ()>(output).encode(&mut buf, &mut ());
+
+                // Now that the response has been serialized, invalidate all
+                // symbols registered with the interner.
+                Symbol::invalidate_all();
```
Comment (eddyb) [inline library/proc_macro/src/bridge/client.rs:417]: This also needs to be done in case of panic (in the `unwrap_or_else` below) and probably even before calling `f` above, just to make sure it starts off empty.

**Review by eddyb — CHANGES_REQUESTED** [review]

File: library/proc_macro/src/lib.rs (hunk 2 above)
Comment (mystor) [reply library/proc_macro/src/lib.rs:1382]: I can move it to be a normal method on `Span` as it actually uses no other parts of `Literal` in its implementation currently, If we need to add in extra parameters for validation in the future, we can add those back in when needed.

File: library/proc_macro/src/bridge/symbol.rs (hunk 5 above)
Comment (mystor) [reply library/proc_macro/src/bridge/symbol.rs:151]: These names were copied almost directly from the rustc interner in `rustc_span`, but I'm happy to change them if desired.
https://github.com/rust-lang/rust/blob/73443a05908b0f8043659295a854295013987876/compiler/rustc_span/src/symbol.rs#L1868-L1869

File: library/proc_macro/src/bridge/symbol.rs (hunk 6 above)
Comment (mystor) [reply library/proc_macro/src/bridge/symbol.rs:155]: Yeah, making this `NonZeroU32` does clean things up a bit, good suggestion. I'm thinking perhaps `sym_base` might be reasonable? I don't love calling it an index, as due to this offset, the `Symbol`s don't really end up containing indexes into the strings array.

File: library/proc_macro/src/bridge/handle.rs (hunk 8 above)
Comment (mystor) [reply library/proc_macro/src/bridge/handle.rs:59]: I'm not sure how valuable it would be in a separate PR given that the types being hashed are presumably simpler without this change, but I suppose we could test it if we wanted to. I've kept that part in a separate commit so it can easily be split off.

I think the simple `Vec` storage might only work for `InternedStore`, and it'd require a thread-local counter and only one server per-thread as well, but that's probably reasonable. It seems unlikely we'd get into a situation where a single client thread is shared by multiple different server threads, which could cause a setup like that to break for server-side interning.

The `Vec` approach with a non-atomic counter feels the most reliable for client-side unique name generation.

File: library/proc_macro/src/bridge/server.rs (hunk 7 above)
Comment (mystor) [reply library/proc_macro/src/bridge/server.rs:71]: The codegen in the `api_tags` module's `rpc_encode_decode!` call will try to define a zero-variant `#[repr(u8)] enum {}` to hold enum variant tags, and adding a `#[repr]` to an enum with no variants is a build error.

https://github.com/rust-lang/rust/blob/73443a05908b0f8043659295a854295013987876/library/proc_macro/src/bridge/mod.rs#L230
https://github.com/rust-lang/rust/blob/73443a05908b0f8043659295a854295013987876/library/proc_macro/src/bridge/rpc.rs#L69

It'd not be too difficult to modify the `rpc_encode_decode!` macro to add a special case for empty enums, though it's not necessary if I end up moving the `normalize_and_validate_ident` method to `Symbol`. I'll go ahead and do that for now.

**Review by eddyb — APPROVED** [review]
LGTM, just a couple nits (sorry for the delay!)

**Comment by mystor** [comment]
Odd, I wonder what changed since https://github.com/rust-lang/rust/pull/98189#issuecomment-1172767345 to make opt builds of `serde-derive` slightly slower.

GIven that we basically will always build macros as debug builds (IIRC) I think that this regression likely isn't super relevant, but perhaps the `Literal::to_string()` impl was made more inlineable or something like that... We still see `cargo` overall build improvements.

**Comment by rylev** [comment]
It seems like there isn't any actionable items to address here, and the improvements outweigh the regressions. I'll mark this as triaged. 

@rustbot label: +perf-regression-triaged
